### PR TITLE
add witness template to lints

### PR DESCRIPTION
**Pull request by suaviloquence**
This adds a new optional field `witness_template` to each lint that can be rendered to make a minimal downstream witness package to show how this change could break downstream code.

Design questions:

- What verbosity level should this be printed for.  Do we want an opt-in `--witness` flag or something similar?
- Similarly, do we need to make witness code for every instance of a lint occurring? Or just the first?
- How can we test the correctness of the witness code?

<details>

<summary Current output: </summary>

```

$ cargo r -- semver-checks --manifest-path new/ --baseline-root old
   Compiling cargo-semver-checks v0.34.0 (/home/m/dev/cargo-semver-checks)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 7.08s
     Running `/home/m/dev/cargo-semver-checks/target/debug/cargo-semver-checks semver-checks --manifest-path new/ --baseline-root old`
     Parsing function_missing v0.1.0 (current)
      Parsed [   0.251s] (current)
     Parsing function_missing v0.1.0 (baseline)
      Parsed [   0.253s] (baseline)
    Checking function_missing v0.1.0 -> v0.1.0 (no change)
     Checked [   0.004s] 81 checks: 80 pass, 1 fail, 0 warn, 0 skip

--- failure function_missing: pub fn removed or renamed ---

Description:
A publicly-visible function cannot be imported by its prior path. A `pub use` may have been removed, or the function itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.34.0/src/lints/function_missing.ron

Failed in:
  function function_missing::will_be_removed_fn, previously in file /home/m/dev/cargo-semver-checks/test_crates/function_missing/old/src/lib.rs:1
note: the following downstream code would break:
use function_missing::will_be_removed_fn;
will_be_removed_fn(...);
  function function_missing::pub_use_removed_fn, previously in file /home/m/dev/cargo-semver-checks/test_crates/function_missing/old/src/lib.rs:4
note: the following downstream code would break:
use function_missing::pub_use_removed_fn;
pub_use_removed_fn(...);
```

</details>

Todo list for me:

- [ ] add a default None field on every existing lint (didn't want to blow up git history just yet, it currently works with `#[serde(default)]`
- [ ] add to `make_new_lint.sh`  - should this be `Some("TODO")` or `None` - not every lint may have a possible witness
- [ ] I was able to put together syntax highlighting using the library portion of `bat` like `cargo-expand` - [branch link](https://github.com/suaviloquence/cargo-semver-checks/compare/witness...suaviloquence:cargo-semver-checks:witness-color).  Do we want this? It adds some dependencies.
- [ ] of course, add more witness templates (this can be a good-first-issue)

### CODE REVIEW

**Comment by obi1kenobi** [comment]
Let's keep `#[serde(default)]` on the new field for now. I think it'll let us be a bit more free to change the representation inside the lints as we evolve how the witnesses get generated.

While generating witness code, we may need to run an additional query to collect contextual information that is only applicable to the witness. For example, say we claim that a function definition has gained an extra argument. The witness generator needs to know:
- whether the function in question is generic
- the names and types of the previous arguments
- whether the function has a return value or not (to avoid a "must use" lint on the witness)

A valid witness may look like this:
```rust
fn witness<T: Debug>(arg_1: i64, arg_2: std::sync::Arc<T>) {
    let _ = broken_function(arg_1, arg_2);
}
```
or it may look like this:
```rust
fn witness() {
    broken_function();
}
```
and we need the machinery to distinguish between the two.

File: src/lints/function_missing.ron
Comment (obi1kenobi) [inline src/lints/function_missing.ron:49]: I think this is a fine starting point, but I wouldn't call this a witness by itself since it won't compile. It's more of a hint toward one. So maybe `hint_template` or something similar would be a more appropriate name?

I think eventually we'll want to expand upon this to make it easier to generate full witnesses that can generate real working code. While we don't have to implement all that today, it would be lovely to have a sense of the broader design within which today's changes should fit.

For example, I'd recommend adding an entire struct here, with the hint being just one of its fields.

**Comment by obi1kenobi** [comment]
Zooming out — I'm very excited to be moving toward making it possible to produce witness code here!

A lot of the rules we check are obscure and complex, and it's not unreasonable to assume that for some of our users, our lints would be the first time they encounter some of those concepts. Saying "your trait isn't object safe" is much more opaque and less user-friendly than saying "look, `dyn YourTrait` won't work anymore."

So I love this, and I'm excited to see it move forward!

File: src/lints/function_missing.ron
Comment (suaviloquence) [reply src/lints/function_missing.ron:49]: > I think this is a fine starting point, but I wouldn't call this a witness by itself since it won't compile. It's more of a hint toward one. So maybe hint_template or something similar would be a more appropriate name?

It's not a witness because it's not an item, so it just needs to be wrapped in a function in this case? What other properties would a witness need to have?

**Comment by suaviloquence** [comment]
Not at computer right now, but something like:

```rust
#[derive(Default)]
struct Witness {
  hint_template: Option<String>,
  witness_template: Option<String>,
  witness_query: Option<Query>
}

struct Query {
  query: String,
  values: BTreeMap<String, String>
}
```

(names notwithstanding and not sure of the value type for query.values). 

where if the query is Some, we run it on the crate diff and provide the values to the witness_template.

File: src/lints/function_missing.ron
Comment (obi1kenobi) [reply src/lints/function_missing.ron:49]: It also elides the arguments to the function by making the function call `will_be_removed_fn(...);`

The tension here is this:
- The full witness is needed since it's the ultimate correctness check.
- But the full witness is usually too onerous and heavyweight to show to users inline with regular CLI output. A rustdoc-style (not wrapped in a function, etc.) _hint_ for how the witness works would be preferable in the CLI.
- The CLI output could also provide text to the effect of "we generated the full witness in this temp file, you can go check it out if you want to verify it's valid."

**Comment by obi1kenobi** [comment]
I think that's a great place to start. I'd rename `values` to match the name in the lint query itself, and change its type to `BTreeMap<String, FieldValue>` to reflect all possible types that could go there. Otherwise, it's great!

**Comment by suaviloquence** [comment]
Updated struct and documented the members.  I used `TransparentValue` instead of `FieldValue` like in `SemverQuery.arguments` so we don't have to specify types in the serialization.

Design questions that came up, particularly about the additional query:

- Should `hint_template` also be provided the additional query? (I think it would be good for efficiency if we could write hints such that they don't need additional parameters, but I don't know if that's possible).
- Should the additional query itself be provided the results of the original SemverQuery? I would say yes, because otherwise we would have to run each query twice when there is a witness query, but this may complicate things e.g., if the witness query does not use all the `@output` directives (I know this is an error if the SemverQueries have unused arguments, but I don't know where this validation occurs)
- How should `@output`s be handled when there is one of the same name in the original query and in the additional query.  It could be an error, or the additional query could take precedence (I wrote the latter in the docs, but we could also do the other one).
- When should we generate witnesses? Should c-s-c show hints by default but opt-in to generate (and test compilation of?) full witness programs?  This is my initial choice, but it would depend on whether we run additional queries for hints, because we don't want to slow down c-s-c by default.

File: src/check_release.rs
Comment (obi1kenobi) [inline src/check_release.rs:161]: Let's make sure the user can't accidentally think we've already *found* some specific downstream code that we're complaining about, and instead we're presenting a hypothetical piece of code that would be broken:
```suggestion
                config.shell_note("downstream code similar to the following would break:")?;
```

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:130]: Couldn't we keep `serde(default)` and make this `Option<Witness>`. Otherwise we might end up in a weird situation where we have no hint text (so nothing to show the user) even though we can generate a full witness example. We'd like weird states like that to be unrepresentable.

I'd make this `Option<Witness>` then make the hint required but the other fields inside still be `Option`.

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:258]: Your doc comments game continues to be excellent 🚀

File: src/check_release.rs
Comment (obi1kenobi) [inline src/check_release.rs:162]: I'm a bit worried about readability here. It's too early to fuss over spacing and the exact CLI output format, but just wanted to flag this for us to revisit as a polish item before shipping this feature.

**Review by obi1kenobi — COMMENTED** [review]
Good call on using `TransparentValue` instead!

> Should hint_template also be provided the additional query? (I think it would be good for efficiency if we could write hints such that they don't need additional parameters, but I don't know if that's possible).

I'd argue "no" here. If the hint needs something else output, the main query should output it.

> Should the additional query itself be provided the results of the original SemverQuery? I would say yes, because otherwise we would have to run each query twice when there is a witness query, but this may complicate things e.g., if the witness query does not use all the @output directives (I know this is an error if the SemverQueries have unused arguments, but I don't know where this validation occurs)

Probably yes. But it might be good to be explicit about which results from the original query it wants.

Perhaps with some serde magic we could specify the query arguments like:
```
{
    "some_var": "custom value",
    "inherited": Inherit("output_from_other_query"),
}
```
What do you think about this approach, assuming it's doable? If you think it's good, let's see if it can be done in reasonable fashion.

> How should @outputs be handled when there is one of the same name in the original query and in the additional query. It could be an error, or the additional query could take precedence (I wrote the latter in the docs, but we could also do the other one).

Probably safer to make it error, I think. I don't see a lot of value in allowing shadowing, whereas it could result in some unfortunate debugging if it happens by accident.

> When should we generate witnesses? Should c-s-c show hints by default but opt-in to generate (and test compilation of?) full witness programs? This is my initial choice, but it would depend on whether we run additional queries for hints, because we don't want to slow down c-s-c by default.

Showing hints by default and generating/testing witnesses when explicitly asked sounds right to me.

File: src/check_release.rs
Comment (suaviloquence) [reply src/check_release.rs:162]: In terms of spacing between items, or the messages themselves (or both?)  I did get syntax highlighting (on the other branch) working for code snippets so that could help for the snippets themselves.

**Comment by suaviloquence** [comment]
Added `InheritedValue` which (should) deserialize `Inherited("...")` as `Inherited(...)` and other values as `TransparentValue`s.  I don't know if the witness query itself should go in this PR, but I can add a test to make sure deserialization works.

I'm renaming the `Query` to `WitnessQuery` because it now only takes `InheritedValue`s, but it is also an option to make it `Query<V>` where `V` is any value type.

File: src/check_release.rs
Comment (obi1kenobi) [reply src/check_release.rs:162]: Spacing, mostly, both between the note and the code, and possible indentation of the hint code. We want to do our best to be readable but concise, and match cargo/rustc style to make merging into cargo easier.

It'll be easier to revisit this later once we have some more witnesses that we're showing. For now, let's leave them hidden behind some permanently-unstable CLI flag (e.g. `-Zunstable-options --show-witness-hints`) so we can iterate freely?

Perhaps adding the infrastructure for such unstable flags would be a good PR in and of itself?

File: src/lints/function_missing.ron
Comment (obi1kenobi) [inline src/lints/function_missing.ron:49]: Could you try `serde(flatten)` in addition to `serde(default)` here? It might make the `Some` unnecessary.

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:256]: We might not need `#[serde(default)]` here since this should always be specified (given the outer option wasn't `None`) I think, right?

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:314]: Excellent, this is exactly what I was hoping for.

**Review by obi1kenobi — COMMENTED** [review]
I like this a lot, well done! 🚀 I think it's safe to start working on a few more witnesses for different lints as stacked PRs, since I don't think our lint file's format will change much from here.

I'm planning to cut a new release next week, and it would be a bit safer if we hid this feature behind a permanently-unstable flag similar to how cargo/rustdoc/rustc etc. hide work-in-progress features. Would you be open to setting up a similar mechanism for us? We'd probably want to mirror their approaches quite closely, down to whether the permanently-unstable flags show up in `--help` output or not.

File: src/query.rs
Comment (suaviloquence) [reply src/query.rs:314]: never mind :sweat_smile: it seems like `ron` has known limitations with `untagged`/`deserialize_any` and having nested `untagged` makes it so the deserialization doesn't work :pensive:.  Maybe we have a list (or map of new_name -> old_name) of `inherit_arguments` along with a map of `arguments`?  Then it's an error if the maps share keys.  Alternately, we could remove the `untagged` so users would have to express arguments as `Inherit("...")` or `Constant(...)`, which prevents duplicate keys from being expressed but is more verbose and less consistent with the `arguments` top-level field.

Comment (obi1kenobi) [reply src/query.rs:314]: Oof, that's unfortunate 😞 What do you think is the better choice between the list/renaming map and the explicit `Inherit(..)` vs `Constant(..)`?

If you feel it's worth writing a few more witness sections before deciding, that's fine by me. We can go the easier route now and revisit based on your experience of using it in the real world.

Comment (suaviloquence) [reply src/query.rs]: I think I would prefer the explicit one.  I also thought of another idea that might work.  It's more jsony than rusty but it does remove the requirement to specify `Constant(...)`:

```ron
arguments: {
  "abc": (inherit: "abc"),
  "string": "literal_string",
  "int": -30,
}
```

where we specify inheriting as a one-field struct `(inherit: "name")` or verbosely, `Inherited(inherit: "name")`.  How do you feel about this.  It won't have any conflicts (I think) with `TransparentValue` because maps and structs in `ron` have different syntax, so `Inherit` and `Constant` variants of the enum should be disjoint.

Comment (suaviloquence) [reply src/query.rs]: I pushed a test that shows that the deserialization is what we expect.

**Comment by obi1kenobi** [comment]
That seems reasonable! Can you just check that lists work too?

File: src/lints/function_missing.ron
Comment (suaviloquence) [reply src/lints/function_missing.ron:49]: I tried to add it and it broke deserialization of the outer `SemverQuery` field :face_with_diagonal_mouth:

**Comment by suaviloquence** [comment]
Added list to test and it works! :+1:

File: src/lints/function_missing.ron
Comment (suaviloquence) [reply src/lints/function_missing.ron:49]: It's possible to use ron's `implicit_some` option to be able to specify it without the `Some`, but this would allow (but not require) this behavior on all `Option<T>` in the `SemverQuery`.  [branch link](https://github.com/suaviloquence/cargo-semver-checks/compare/witness...implicit-some)

Comment (obi1kenobi) [reply src/lints/function_missing.ron:49]: Oh neat, TIL about that extension. I think that's fine, yeah. Maybe good for a separate PR we can merge immediately?

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:802]: Heterogeneous lists like this aren't something Trustfall supports. This test case might make future refactoring harder if e.g. the `TransparentValue` type changes in a way that disallows that at the type level — the person doing it will have to also fix this test case too. It's usually not a good idea to exceed the bounds of the spec, because if the implementation of the spec starts enforcing the bounds more strongly then breakage ensues.

I'd recommend removing the string here, and making a separate list entry with a string or two to test strings inside the list.

**Comment by obi1kenobi** [comment]
All looks good other than the tiny nitpick on the test structure!

**Comment by suaviloquence** [comment]
Rebased on #897 and fixed test.

**Comment by suaviloquence** [comment]
> Btw, is the failing test accurate in saying that we print the error template to stdout and the hint to stderr? That seems strange and probably not desirable. Might be time to look into getting our use of stdout/stderr to be more consistent, or at least maybe introducing the ability to print notes to either stdout/stderr as needed in the caller's context.

I think any of the `shell_*` only prints to stderr.  AFAIK, it's just the lint output that goes to stdout.  Also, should the tests have witness hints while it's behind a #896 feature gate?  That's why I haven't updated the outputs yet.

**Comment by obi1kenobi** [comment]
> I think any of the `shell_*` only prints to stderr. AFAIK, it's just the lint output that goes to stdout.

I think this is the first time we've used `note:` styled output on something that's relevant to stdout. Previously it was just for telling people that c-s-c failed to run for some reason, for which stderr was probably the right output.

I think it's confusing to direct the witness to stderr while the lint output is on stdout. I think the witness info should be on stdout too, so the user doesn't have to figure out how to interleave the stdout/stderr outputs together to figure out what each witness was referring to.

> Also, should the tests have witness hints while it's behind a #896 feature gate? That's why I haven't updated the outputs yet.

Probably a good idea to test with the witnesses present, even though it's an unstable feature. We still want to know if we broke something, or else we'll have a really hard time stabilizing it.

**Comment by suaviloquence** [comment]
Added configuration to indicate in a `Check` whether to print witnesses (the `WitnessGeneration` struct).

Todos:

- [ ] make the note print to stdout (might have to manually impl it)
- [ ] set up some sort of test harness per lint that checks if witness is `Some`, then loads a `test_outputs/witnesses/{query_name}.ron` and generates witness code/hints and compares expected == actual

**Comment by suaviloquence** [comment]
Added test harness and moved note to stdout, so this should be ready for review.

File: test_outputs/witnesses/function_missing.output.ron
Comment (suaviloquence) [inline test_outputs/witnesses/function_missing.output.ron:6]: it's mentioned in the comment in the test code, but prettier (multiline) strings are blocked on the release of ron v0.9 (which although there is an alpha, doesn't seem to be happening anytime soon due to maintainer overburden).  it may be worth (albeit hacky), doing a `str.replace("\\n", "\n")` after serializing to a string for better diffs.

File: src/query.rs
Comment (suaviloquence) [inline src/query.rs:730]: once we have the infrastructure to run witness queries and render full witnesses, code here should call an encapsulated version of that logic to generate the full witness here, if it is `Some`.

File: src/check_release.rs
Comment (suaviloquence) [inline src/check_release.rs:168]: here i just inlined the note code.  we can't have a method in `GlobalConfig` that takes an `impl Write` and renders a note/error/warning/etc., because `.stdout`/`.stderr` mutably borrows the whole `GlobalConfig` struct.  we could:
- make (associated) functions
- implement them on the streams themselves (using a newtype)
- revisit using a crate like `annotate-snippets` to render for us
- make macros

if we want a longer-term solution, if you think it is necessary.

**Comment by suaviloquence** [comment]
@obi1kenobi btw this one is ready for review when you get a chance.

File: src/check_release.rs
Comment (obi1kenobi) [reply src/check_release.rs:168]: Revisiting with something like `annotate-snippets` down the line would be great. This is a reasonable solution for the meantime — it doesn't make sense to hold up witness generation over CLI formatting.

File: src/lib.rs
Comment (obi1kenobi) [inline src/lib.rs:48]:
```suggestion
    /// Options for generating [witnesses](Witness).
```

File: src/main.rs
Comment (obi1kenobi) [reply src/main.rs:631]: I don't have strong feelings about it, this is fine for now.

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:257]:
```suggestion
    /// show the idea of why downstream code could break.  It will be provided all
    /// `@output` data from the [`SemverQuery`] query that contains this [`Witness`].
```

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:265]:
```suggestion
    /// use {{join "::" path}};
    /// {{name}}(...);
```

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:292]: This is a very reasonable but unfortunately incorrect witness. This stuff is hard!

The issue is that `name` is the _declared_ name of the enum, but it might be renamed as part of the importable path. For example:
```rust
mod internal {
    pub enum Example {
        Variant,
    }
}

pub use internal::Example as PubExample;
```

If we removed `Variant` here, the generated witness will say:
```rust
use my_crate::PubExample;
fn witness(item: Example) {
    if let Example::Variant{..} = item {

    }
}
```
This won't compile since `Example` is undefined.

The easiest fix is to always use the fully-qualified path instead of importing it.
```suggestion
    /// # let _ = r#"
    /// fn witness(item: {{path}}) {
    ///     if let {{path}}::{{variant_name}} {..} = item {
    ///
    ///     }
    /// }
    /// # "#;
```

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:704]: In a separate PR, could you amend our contributing docs for adding new lints to add an optional step for "add a witness" and add a small section explaining how to do it?

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:741]: Marking this TODO to make it stand out a bit more.
```suggestion
                // TODO: ideally we would use the `escape_strings` feature to pretty-print
                //       multiline strings, but this is blocked on ron releasing v0.9:
```

**Review by obi1kenobi — APPROVED** [review]
Looks good! Making some trivial tweaks and merging. Also leaving an action item for you for a subsequent PR on updating contributing docs to explain the witness section.

File: src/query.rs
Comment (obi1kenobi) [inline src/query.rs:726]:
```suggestion
                                // TODO: Run witness queries and generate full witness here.
                                WitnessOutput {
```

File: src/query.rs
Comment (obi1kenobi) [reply src/query.rs:730]: Added a TODO note to our future selves about it.

**Comment by obi1kenobi** [comment]
> I was able to put together syntax highlighting using the library portion of bat like cargo-expand - [branch link](https://github.com/suaviloquence/cargo-semver-checks/compare/witness...suaviloquence:cargo-semver-checks:witness-color). Do we want this? It adds some dependencies.

I'd love to see a before-and-after screenshot of the difference if you get a chance! I'm open to adding a dependency if it results in notable improvement in user experience, and syntax highlighting certainly could be it.

**Comment by obi1kenobi** [comment]
Ah unfortunately this has a merge conflict due to the priority logic for lints that merged yesterday. Sorry about that!

**Review by obi1kenobi — APPROVED** [review]

**Pull request merged**
